libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt

```diff
@@ -0,0 +1,163 @@
+        val snappingResultList = directionsRoute.getSnappingResultList()
+
+        var index = -1
+        directionsRoute.legs()?.forEachIndexed { routeLegIndex, routeLeg ->
+            index++
+
+            val legLastGeometryIndex = directionsRoute.stepsGeometryToPoints(routeLeg).lastIndex
+
+            val silentWaypoints = routeLeg.silentWaypoints()
+
+            routeLeg.closures()?.forEach { closure ->
+                val silentWaypointsInClosureRange = silentWaypoints.inGeometryRange(
+                    closure.geometryIndexStart(), closure.geometryIndexEnd()
+                )
+
+                if (closure.geometryIndexStart() == 0 &&
+                    snappingResultList.getOrNull(index) != true
+                ) {
+                    logD(
+                        "Route with id [${this@hasUnexpectedClosures.id}] has closure at the " +
+                            "start of the leg, leg index $routeLegIndex",
+                        LOG_CATEGORY
+                    )
+                    return@withContext true
+                }
+                if (closure.geometryIndexEnd() == legLastGeometryIndex &&
+                    snappingResultList.getOrNull(index + silentWaypoints.size + 1) != true
+                ) {
+                    logD(
+                        "Route with id [${this@hasUnexpectedClosures.id}] has closure at " +
+                            "the end of the leg, leg index $routeLegIndex",
+                        LOG_CATEGORY
+                    )
+                    return@withContext true
+                }
+                if (silentWaypointsInClosureRange.isNotEmpty()) {
+                    silentWaypointsInClosureRange.forEach {
+                        val isSnapAllowed =
+                            snappingResultList.getOrNull(index + 1 + it) ?: false
+                        if (!isSnapAllowed) {
+                            logD(
+                                "Route with id [${this@hasUnexpectedClosures.id}] has closure " +
+                                    "at leg index $routeLegIndex, that overlaps silent (via) " +
+                                    "waypoint",
+                                LOG_CATEGORY
+                            )
+                            return@withContext true
+                        }
+                    }
+                }
+                if (closure.geometryIndexStart() != 0 &&
+                    closure.geometryIndexEnd() != legLastGeometryIndex &&
+                    silentWaypointsInClosureRange.isEmpty()
+                ) {
+                    logD(
+                        "Route with id [${this@hasUnexpectedClosures.id}] has closure at leg " +
+                            "index $routeLegIndex",
+                    )
+                    return@withContext true
+                }
+            }
+            index += silentWaypoints.size
+        }
+
+        return@withContext false
```

Comment:
This should be flipped.

```suggestion
        return@withContext true
```

The goal is to do as little checks as possible to confirm that a closure is allowed, and only when we don't find any acceptable criteria return `true`.

This allows us to, for example, first check for the silent waypoints (which is the cheapest check because it doesn't require deserialization of geometries), and only check for all regular waypoints, and ultimately fail if no criteria was found.

Reply:
must be returned `false`: it means that the route `doesn't` have unexpected closures.

>This allows us to, for example, first check for the silent waypoints (which is the cheapest check because it doesn't require deserialization of geometries), and only check for all regular waypoints, and ultimately fail if no criteria was found.

we have to measure geometry before the `for loop`, so geometry still is measured, event if we check silent waypoints first.